### ISSUE TITLE
Support Promises

### ISSUE THREAD

**Issue by dmihal** [ethereumjs/ethereumjs-vm#252]
I would love if the API supported returning Promises as an alternative to callbacks. This would be implemented by returning Promises if no callback is provided.

I'm happy to add the feature and submit a PR if you're open to it!

**Comment by jwasinger** [comment]
I like this idea.  It is backwards compatible as well!

Pending any reservations by @holgerd77 @axic , I would say go ahead :)

**Comment by holgerd77** [comment]
Would also very much support this, great that you want to implement it! 😄 

Could you post some code snippet with 10LOC or so on how you think/plan this would look like when being called with promises from the outside?

**Comment by dmihal** [comment]
With promises, the code can be executed in 3 ways:

The existing way with callbacks
```javascript
vm.runCode({
  code: Buffer.from(code, 'hex'), // code needs to be a Buffer
  gasLimit: Buffer.from('ffffffff', 'hex')
}, function(err, results){
  console.log('returned: ' + results.return.toString('hex'));
})
```

Using promises
```javascript
vm.runCode({
  code: Buffer.from(code, 'hex'), // code needs to be a Buffer
  gasLimit: Buffer.from('ffffffff', 'hex')
})
.then(function(results){
  console.log('returned: ' + results.return.toString('hex'));
})
.catch(function(err){
  console.error(err);
})
```

Or using async/await

```javascript
try {
  var results = await vm.runCode({
    code: Buffer.from(code, 'hex'), // code needs to be a Buffer
    gasLimit: Buffer.from('ffffffff', 'hex')
  });
  console.log('returned: ' + results.return.toString('hex'));
} catch (err) {
  console.error(err);
}
```

**Comment by dmihal** [comment]
Based on the feedback, I made a cleaner implementation in #273

**Comment by Silur** [comment]
@jwasinger actually I made a Promised (no async dep) version of all etherumjs repos with @wanderer in around last year may

**Comment by holgerd77** [comment]
@Silur Tell us more, is this code locked on an USB-Stick in a secret cupboard in a forgotten Lisbon library?? 😄 🔑

**Comment by holgerd77** [comment]
Addressed in #546, will close.
